presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/OrcStorageManager.java

```diff
@@ -0,0 +1,575 @@
+                return completedFuture(ImmutableList.of());
+            }
+            return supplyAsync(
+                    () -> rewriteChunk(transactionId, tableId, bucketNumber, chunkId, chunkColumnTypes, rowsToDelete),
+                    deletionExecutor);
+        };
+    }
+
+    private void writeChunk(long chunkId)
+    {
+        if (!chunkStore.chunkExists(chunkId)) {
+            throw new PrestoException(RAPTOR_CHUNKSTORE_ERROR, "Chunk does not exist in chunk store after put");
+        }
+
+        File stagingFile = storageService.getStagingFile(chunkId);
+        File storageFile = storageService.getStorageFile(chunkId);
+
+        try {
+            Files.move(stagingFile.toPath(), storageFile.toPath(), ATOMIC_MOVE);
+        }
+        catch (IOException e) {
+            throw new PrestoException(RAPTOR_STORAGE_ERROR, "Failed to move chunk file", e);
+        }
+    }
+
+    private OrcDataSource openChunk(long tableId, long chunkId, ReaderAttributes readerAttributes)
+    {
+        File file = storageService.getStorageFile(chunkId).getAbsoluteFile();
+
+        if (!file.exists()) {
+            try {
+                Future<?> future = recoveryManager.recoverChunk(tableId, chunkId);
+                future.get(recoveryTimeout.toMillis(), TimeUnit.MILLISECONDS);
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            }
+            catch (ExecutionException e) {
+                throwIfInstanceOf(e.getCause(), PrestoException.class);
+                throw new PrestoException(RAPTOR_RECOVERY_ERROR, "Error recovering chunk: " + chunkId, e.getCause());
+            }
+            catch (TimeoutException e) {
+                throw new PrestoException(RAPTOR_RECOVERY_TIMEOUT, "Chunk is being recovered from chunk store. Please retry in a few minutes: " + chunkId);
+            }
+        }
+
+        try {
+            return fileOrcDataSource(file, readerAttributes);
+        }
+        catch (IOException e) {
+            throw new PrestoException(RAPTOR_STORAGE_ERROR, "Failed to open chunk file: " + file, e);
+        }
+    }
+
+    private ChunkInfo createChunkInfo(long chunkId, int bucketNumber, File file, long rowCount, long uncompressedSize)
+    {
+        return new ChunkInfo(chunkId, bucketNumber, computeChunkStats(file), rowCount, file.length(), uncompressedSize, xxhash64(file));
+    }
+
+    private List<ColumnStats> computeChunkStats(File file)
+    {
+        try (OrcDataSource dataSource = fileOrcDataSource(file, defaultReaderAttributes)) {
+            OrcReader reader = createOrcReader(dataSource, defaultReaderAttributes);
+
+            ImmutableList.Builder<ColumnStats> list = ImmutableList.builder();
+            for (Entry<Long, Type> entry : getColumnTypes(reader).entrySet()) {
+                computeColumnStats(reader, entry.getKey(), entry.getValue()).ifPresent(list::add);
```

Comment:
I know this is a legacy. Not saying we should change now. Just wanna mark the issue: it may not be necessary to read all data back just for the min max. 

Will it have the same effect as "ORC Writer validation"?

Reply:
Agreed, collecting min/max during the write would avoid the cost of reading back the data (decompression and other cost). This could be a good area of future optimization if writing is shown to be a significant cost of the overall system.